src/AWSCredentials.jl

```diff
@@ -56,392 +58,460 @@ mutable struct AWSCredentials
-        return true
-    end
+            new_aws_creds === nothing && error("Can't find AWS credentials!")
+            copyto!(aws_creds, new_aws_creds)
 
-    if isfile("/sys/devices/virtual/dmi/id/product_uuid")
-        try
-            # product_uuid is not world readable!
-            # https://patchwork.kernel.org/patch/6461521/
-            # https://github.com/JuliaCloud/AWSCore.jl/issues/24
-            if String(read("/sys/devices/virtual/dmi/id/product_uuid")) == "EC2"
-                return true
-            end
-        catch
+            # Ensure credential_method is not overwritten by the new credentials
+            aws_creds.renew = credential_method
         end
     end
 
-    return false
+    return aws_creds
 end
 
-localhost_maybe_ec2() = localhost_is_ec2() ||
-                        isfile("/sys/devices/virtual/dmi/id/product_uuid")
-
-"""
-    aws_user_arn(::AWSConfig)
-
-Unique
-[Amazon Resource Name]
-(http://docs.aws.amazon.com/IAM/latest/UserGuide/id_users.html)
-for configrued user.
 
-e.g. `"arn:aws:iam::account-ID-without-hyphens:user/Bob"`
-"""
-function aws_user_arn(aws::AWSConfig)
-    creds = aws[:creds]
-    if creds.user_arn == ""
-        r = Services.sts(aws, "GetCallerIdentity", [])
-        creds.user_arn = r["Arn"]
-        creds.account_number = r["Account"]
-    end
-    return creds.user_arn
+function _will_expire(aws_creds::AWSCredentials)
+    return now(UTC) >= aws_creds.expiry - Minute(5)
 end
 
 
 """
-    aws_account_number(::AWSConfig)
-
-12-digit [AWS Account Number](http://docs.aws.amazon.com/general/latest/gr/acct-identifiers.html).
-"""
-function aws_account_number(aws::AWSConfig)
-    creds = aws[:creds]
-    if creds.account_number == ""
-        aws_user_arn(aws)
-    end
-    return creds.account_number
-end
+    _ec2_metadata(metadata_endpoint::String) -> Union{String, Nothing}
 
+Retrieve the EC2 meta data from the local AWS endpoint. Return the EC2 metadata request
+body, or `nothing` if not running on an EC2 instance.
 
-"""
-    ec2_metadata(key)
+# Arguments
+- `metadata_endpoint::String`: AWS internal meta data endpoint to hit
 
-Fetch [EC2 meta-data]
-(http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-instance-metadata.html)
-for `key`.
+# Throws
+- `StatusError`: If the response status is >= 300
+- `ParsingError`: Invalid HTTP request target
 """
-function ec2_metadata(key)
+function _ec2_metadata(metadata_endpoint::String)
+    try
+        request = @mock HTTP.request(
+            "GET",
+            "http://169.254.169.254/latest/meta-data/$metadata_endpoint",
+            retry=false,
+            retries=0,
+            readtimeout=5
+        )
 
-    @assert localhost_maybe_ec2()
+        return String(request.body)
+    catch e
+        e isa IOError || rethrow(e)
+    end
 
-    String(http_get("http://169.254.169.254/latest/meta-data/$key").body)
+    return nothing
 end
 
-function instance_credentials()
+
+function _get_awscreds_instance()
```

Comment:
`_get_awscreds_instance()`
`ec2_instance_credentials()`
`get_awscreds_env_vars()`

These are inconsistent in name and underscoredness and should be changed or clarified